README.md
# akt - ancestry and kinship toolkit
See http://illumina.github.io/akt/index.html for documentation.

##License

AKT is freely available under the [GPL3 license](https://github.com/Illumina/agg/blob/master/LICENSE). AKT is not commercially supported.

AKT relies on HTSlib and Eigen. [Eigen](http://eigen.tuxfamily.org/index.php?title=Main_Page) is a header-only library for matrix algebra released under the [MPL2 license](https://www.mozilla.org/en-US/MPL/2.0/). [HTSlib](http://www.htslib.org/) is a library for efficently parsing vcf/bcf files released under the [MIT/Expat License](http://choosealicense.com/licenses/mit/).

##Installation instructions

```
git clone https://github.com/Illumina/akt.git
cd akt/
make
```
If you get a warning about 'omp.h not found' (especially on Macs)
you can try
```
make no_omp
```
Everything will be run on a single thread, so the `-n` option does nothing
in `akt kin` and `akt ibd`.

##Using akt
akt uses the syntax
```
./akt COMMAND <options>
```
To see a list of available tools use
```

Program:	akt (Ancestry and Kinship Tools)
Version:	48e66c0
Contact:	rarthur@illumina.com

Copyright (c) 2016, Illumina, Inc. All rights reserved. See LICENSE for further details.

Usage:	akt <command> [options]

	pca                      principal component analysis
	kin                      detect average IBD sharing
	relatives                discover pedigrees
	ibd                      detect segments shared IBD
	mendel                   profile Mendelian inhertiance and inconsistencies in known pedigrees
	cluster                  perform cluster analyses
	LDplot                   output correlation matrix
	stats                    calculate AF and LD metrics
	metafreq                 examine two files for AF differences

```
##kin
```
Calculate IBD stats from a VCF
Usage:
./akt kin in.bcf -T sites.vcf.gz
Expects input.bcf to contain genotypes.
User must specify a file containing allele frequencies with -R!
	 -k --minkin:			threshold for relatedness output (none)
	 -u --unnorm:			If present don't normalize
	 -c --calc:			calculate AF from data
	 -h --thin:			keep every t variants
	 -T --targets-file:		intersecting VCF
	 -n --nthreads: 		num threads
	 -a --aftag:			allele frequency tag
	 -r --regions:			chromosome region
	 -R --regions-file:		list of regions, file
	 -s --samples:			list of samples
	 -S --samples-file:		list of samples, file
	 -f --pairfile:			file containing sample pairs
	 -m --maf:			minimum MAF

```
* -k : Output only pairs with kinship coefficient more than this.
* -u : Don't normalize IBD scores
* -c : Calculate allele frequency from data
* -h : Hop this many markers forward from last accepted marker. 
* -T : Indexed VCF file containing intersecting sites and relevant site info. 
* -n : Number of threads to use. 
* -a : Allele frequency tag e.g. "TAG" reads or writes "TAG_AF" in the target VCF. 
* -r : Comma-separated list of regions, chr:from-to.
* -R : File containing 3 columns: CHROM, POS and POS_TO. 
* -s : Comma-separated list of samples to include or exclude if prefixed with "^".
* -S : File of sample names to include or exclude if prefixed with "^". One sample per line.
* -f : File containing pairs of samples to compute, 2 column: Sample1 Sample2. 
* -m : Minimum MAF a site to be counted. When reading: if AF > 0.5 checks 1-AF > m. 


Run the kinship calculation by giving akt a multi-sample vcf/bcf file:
```
~/ancestry_tools$./akt kin multisample.bcf -R data/1000G.snps.nochr.vcf.gz -n 32 > allibd
```
the output `allibd` is 7 columns in the format
```
ID1 ID2 IBD0 IBD1 IBD2 KINSHIP NSNP
```
The algorithm used to calculate IBD is taken from [PLINK](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1950838/) with some minor changes:
* No 'bias correction' since allele frequencies are assumed to be accurate
* Normalization as follows:
	* if IBD0 > 1: IBD0 = 1, IBD1 = 0, IBD2 = 0
	* if IBD1 < 0: IBD1 = 0
	* if IBD2 < 0: IBD2 = 0
	* norm = IBD0 + IBD1 + IBD2
	* IBD0 /= norm, IBD1 /= norm; IBD2 /= norm;
* We do NOT follow PLINK which forces IBD to obey consistency conditions - this affects the clustering that is required for the `relatives` code.
* KINSHIP = 0.5 * IBD2 + 0.25 * IBD1;

##relatives
```
Discover relatives from IBD
Usage:
./akt relatives ibdfile
	 -k --kmin:			threshold for relatedness (0.05)
	 -i --its:			number of iterations to find unrelated (10)
	 -g --graphout:			if present output pedigree graph files
	 -p --prefix:			output file prefix (out)
arrow types     : solid black	= parent-child
                : dotted black	= siblings
                : blue 		= second order
                : red		= duplicates
                : directed	= from parent to child

```
* -k : Only keep links with kinship above this threshold (searches in this set for duplicate, parent-child and sibling links).
* -i : Iteration parameter for unrelated set output.
* -p : Prefix for output files.
* -g : If present output graphviz files. These can be visulaised using e.g. `neato -Tpng -O out.allgraph` or for family pedigrees `dot -Tpng -O out.Fam0.graph`.
```
./akt relatives allibd -g > allrelatives
```
The output contains duplicates, families, relationship types and unrelated individuals
```
grep ^Dup allrelatives
Dup0 Sample0
Dup0 Sample1
...
grep ^Fam allrelatives
Fam0 Sample2
Fam0 Sample3
...
...
grep ^Type allrelatives
Type Fam0 Sample2 Sample3 Parent/Child
...
grep ^Unrel allrelatives
Sample0
Sample2
...
```
The file `out.allgraph` can be viewed with gviz e.g. `fdp out.allgraph -Tpng -O` and the families can be viewed using 
e.g. `dot out.Fam0.graph -Tpng -O`. The parent child relations are also recorded in PLINK fam format in `out.fam`. If
e.g. a sibling pair, is found the samples will appear in `out.fam` without parents. If the direction of the relationship
can't be determined e.g. for parent/child duos a random sample is assigned to be the parent in `out.fam`. The final column
in the `.fam` file specifies how many potential parents the sample had.

##mendel
```

About:   akt mendel - profiles duo/trios
Usage:   ./akt mendel input.bcf -p pedigree.fam

Options:
    -p, --pedigree              pedigree information in plink .fam format
    -o, --out                   output a site only vcf.gz annotated with site specific error rates
    -i, --include               variant filters to apply eg. -i 'TYPE==snp && QUAL>=10 && DP<100000 && HWE<10' 
    -t, --targets [^]<region>   Set regions. Exclude regions with "^" prefix
    -r, --regions <region>      restrict to comma-separated list of regions
    -R, --regions-file <file>   restrict to regions listed in a file
```
* -o : Site only vcf annotated with mendel error rates
* -i : Apply filters see [bcftools expressions](https://samtools.github.io/bcftools/bcftools.html#expressions)
* -t : Regions/sites to include/exclude see [bcftools](https://samtools.github.io/bcftools/bcftools.html#common_options)
* -r : Comma-separated list of regions, chr:from-to.
* -R : File containing 3 columns: CHROM, POS and POS_TO. 
* -p : The `-p` option specifies the pedigree structure and uses the well-known six column plink .fam format:

1. Family ID
2. Individual ID
3. Paternal ID
4. Maternal ID
5. Sex (1=male; 2=female; other=unknown)
6. Phenotype
Here is an example pedigree file for the [CEPH1463](http://www.nature.com/nmeth/journal/v12/n10/fig_tab/nmeth.3505_SF4.html) pedigree:

```
PG NA12889 0 0  1 -9
PG NA12890 0 0  2 -9
PG NA12891 0 0  1 -9
PG NA12892 0 0  2 -9
PG NA12877 NA12889 NA12890      1 -9
PG NA12878 NA12891 NA12892      2 -9
PG NA12879 NA12877 NA12878      2 -9
PG NA12880 NA12877 NA12878      2 -9
PG NA12881 NA12877 NA12878      2 -9
PG NA12882 NA12877 NA12878      1 -9
PG NA12883 NA12877 NA12878      1 -9
PG NA12884 NA12877 NA12878      1 -9
PG NA12885 NA12877 NA12878      2 -9
PG NA12886 NA12877 NA12878      1 -9
PG NA12887 NA12877 NA12878      2 -9
PG NA12888 NA12877 NA12878      1 -9
PG NA12893 NA12877 NA12878      1 -9
```
##pca
```
Performs principal component analysis on a vcf/bcf
Usage:
./akt pca input.bcf
	 -T --targets-file:		intersecting VCF
	 -o --output:			output vcf
	 -O --outputfmt:		output vcf format
	 -r --regions:			chromosome region
	 -R --regions-file:		list of regions, file
	 -s --samples:			list of samples
	 -S --samples-file:		list of samples, file
	 -h --thin:			keep every t variants
	 -m --maf:			minimum MAF
	 -w --weight:			VCF with weights for PCA
	 -N --npca:			first N principle components
	 -a --alg:			exact SVD (slow)
	 -C --covdef:			definition of SVD matrix: 0=(G-mu) 1=(G-mu)/sqrt(p(1-p)) 2=diag-G(2-G) default(1)
	 -e --extra:			extra vectors for Red SVD
	 -F --svfile:			File containing singular values

```
* -T : Indexed VCF file containing intersecting sites and relevant site info. 
* -o : Output to vcf.
* -O : Output format of vcf b=compressed bcf, z=compressed vcf.
* -r : Comma-separated list of regions, chr:from-to.
* -R : File containing 3 columns: CHROM, POS and POS_TO. 
* -s : Comma-separated list of samples to include or exclude if prefixed with "^".
* -S : File of sample names to include or exclude if prefixed with "^". One sample per line.
* -h : Hop this many markers forward from last accepted marker. 
* -m : Minimum MAF a site to be counted. When reading: if AF > 0.5 checks 1-AF > m. 
* -w : Use precalculated principle components.
* -N : Number of principle components to calculate.
* -a : Use JacobiSVD PCA algorithm, which is exact to float precision but very slow.
* -e : Default PCA calculation is the inexact `RedSVD` algorithm, which requires this parameter. The higher the number the more accurate principle components will be obtained. 
* -F : File to output the singular values.
* -C : Which matrix to take the PCA of. 0 uses mean subtracted genotype matrix; 1 uses mean subtracted and normalized genotype
matrix; 2 uses normalized covariance matrix with bias term subtracted from diagonal elements.  

`pca` can either calculate principle components from data and do projections onto these components or use precalculated 
principle component vectors
```
~/ancestry_tools$ ./akt pca multisample.bcf -O b -o pca.bcf -m 0.05 -h 10 -N 5 > projections 
```
The file `projections` contains
```
SAMPLE_ID0 P0 P1 P2 P3 P4
SAMPLE_ID1 P0 P1 P2 P3 P4
...
```
The bcf file `pca.bcf` contains
```
bcftools query -f "%INFO/WEIGHT\n" pca.bcf
pc00 pc01 pc02 pc03 pc04
pc10 pc11 pc12 pc13 pc14
...
```
First index is the site index and second is which principle component. If you already have precalculated PCA weights and 
you want to project new samples onto those weights
```
./akt pca new_multisample.bcf -w pca.bcf > projections
```
##cluster
```
Clustering on text files
Usage:   ./akt cluster input.txt
	 -k --K:			number of clusters
	 -i --seed:			random seed for starting values
	 -a --alg:			clustering algorithm 0 = k++means, 1 = gaussian mixture, 2 = density method
	 -c --cols:			column range to read
	 -C --cfile:			initial guess for cluster centres in text file
	 -o --outputcfile:		assigned cluster centres
	 -c --cols:			which columns to use e.g. 2-4
	 -I --maxits:			max number of iterations to use: a=0,1
	 -d --dc:			radius for density method: a=2
	 -p --rho_min:			min density for cluster centre: a=2
	 -D --delta_min:		min radius for cluster centre: a=2
	 --density-plot:		plot the density and finish: a=2
	 -e --silhouette:		calculate silhouette score
```
* -k : Number of clusters. Examine the data to guess this or analyse silhouette scores.
* -i : Random seed.
* -a : 0 to use k++ means clustering, 1 to use EM with Gaussians, 2 to use density method.
* -C : file with initial guess for cluster centres. If analysing 2d data this should contain 2 columns and K rows.
* -c : Which columns in input file to use.
* -I : Maximum number of iterations to use for alg 0 or 1.
* -d : radius around each point for counting density.
* -p : min density for cluster centre.
* -D : min radius for cluster centre.
* --density-plot : plot delta-density graph and finish.
* -e : Calculate silhouette score (goodness of cluster assignment).

e.g. cluster the first three principle components from the output of vcfpca using k++ means. 
Different clusters should correspond to different ancestries
```
./akt cluster projections -k 4 -c 2-4 -e > clustered
```
Each block in `clustered` contains all the data for points in that cluster with the row layout
```
P2 P3 P4 (SILHOUETTE) CLUSTERID P1 P5 ...
```
To visually inspect the clustered data try the following gnuplot command
```
gnuplot> splot for [i=0:3:1] 'clustered' every:::i::i
```
The silhouette column can be used to filter on badly clustered samples. 
~1 means a sample is closer to the centroid of its assigned cluster than the next nearest cluster.
A sample can be badly clustered for many reasons: bad choice of K, convergence of the clustering algorithm 
to bad local minimum, individual is of mixed ancestry. You could filter out badly clustered samples using
```
awk '{ if($4 > 0.5) print $0}' clustered > well_clustered
```
###density clustering
This is a better clustering algorithm but requires some tuning by hand. First run
```
./akt cluster projections -c 2-4 -a 2 -d 1 --density-plot > dplot
```
the file dplot is a two column data file of density versus delta 
(the distance to the closest data point of higher density). By examining this plot outliers 
are obvious and a cutoff in density and delta can be determined that isolates these points. 
Once this cutoff is determined run
```
./akt cluster projections -c 2-4 -a 2 -d 1 -p 0 -D 75 > clustered
```
Where 0 and 75 are the cutoffs determined from `dplot`. This will output the number of clusters that were found 
(including one for unassigned points) and `clustered` will contain the clustered data in the usual format.
```
gnuplot> splot for [i=0:4:1] 'clustered' every:::i::i
```
Unassigned data points are always put in cluster 0. The silhouette score for unassigned data is undefined.

##stats
```
Calculate correlation across a population
Usage:
./akt stats input_filename.vcf
Expects input_filename.vcf to contain hard genotypes
	 -F --flank:			size of left and right flanking region (1000bp)
	 -b --block:			-F argument is number of markers instead of number of base pairs
	 -x --afonly:			calculate allele freq only
	 -c --output_cor:		output sitewise correlation (false)
	 -C --output_cormin:		output sitewise correlation greater than (0)
	 -a --aftag:			allele frequency tag
	 -o --output:			output vcf
	 -O --outputfmt:		output vcf format
	 -r --regions:			chromosome region
	 -R --regions-file:		list of regions, file
	 -s --samples:			list of samples
	 -S --samples-file:		list of samples, file
```
* -F : Correlation with variants in a window of size f base pairs to left and right or each variant. 
* -b : The number in the -F argument now interpreted as number of flanking variants instead of flanking positions.
* -x : Do not calculate covariance matrix.
* -c : Output all correlation.
* -C : Output correlation values greater than this.
* -a : Allele frequency tag e.g. "TAG" reads or writes "TAG_AF" in the target VCF. 
* -o : Output to vcf.

Changelog
#2016.06.03
* kin command defaults to calculating allele frequency from data
* changed allele frequency vcf argument to -F


#2016.05.06
* added a check for 0 samples in akt pca

#2016.04.08
* initial public release



ibd.cpp
#define __STDC_LIMIT_MACROS

#include "akt.hpp"
#include "logs.hpp"
#include "kin.hpp"

using namespace std;

//log odds ratio of IBD versus no IBD for observed genotypes and AF
float calc_snp_logodds(vector< vector<int> > &G, vector< float > &af, vector< float > &ld, int ip, int j1, int j2, vector< vector<float> > &GQ){
		
		//GLs
		float gl1[3]; for(int g=0; g<3; ++g){ gl1[g] = ( g == G[ip][j1] ) ? 1.0-GQ[ip][j1] : GQ[ip][j1]/2.0; } 
		float gl2[3]; for(int g=0; g<3; ++g){ gl2[g] = ( g == G[ip][j2] ) ? 1.0-GQ[ip][j2] : GQ[ip][j2]/2.0; } 

		//no IBD -> HWE
		float den = log( 
				gl1[0] * (1 - af[ip]) * (1 - af[ip])
			+	gl1[1] * (1 - af[ip]) * af[ip] * 2
			+	gl1[2] * af[ip] * af[ip]
			);
		//IBD
		float num = log(
			gl2[0] * ( gl1[0]*(1 - af[ip]) + gl1[1]*af[ip] )
		+	gl2[1] * ( gl1[0]*(0.5*(1 - af[ip])) + gl1[1]*0.5 + gl1[2]*0.5*af[ip] )
		+	gl2[2] * ( gl1[1]*(1 - af[ip]) + gl1[2]*af[ip] )
		);
		
		return (num - den)/ld[ip];
}

/**
 * @name    usage
 * @brief   print out options
 *
 * List of input options
 *
 */
static void usage(){
  cerr << "Find IBD regions from a VCF" << endl;	
  cerr << "Usage:" << endl;
  cerr << "./akt ibd in.bcf -p sites.bcf" << endl;
  cerr << "Expects input_filename.vcf to contain hard genotypes" << endl;
  umessage('T');
  umessage('r');
  umessage('R');
  umessage('s');
  umessage('S');
  umessage('h');
  umessage('a');
  umessage('f');
  umessage('n');
  umessage('m');
  cerr << "\t -e --error: 			If no GQ then this is error probability(1e-3)" << endl;	 
  cerr << "\t -M --thresh: 			likelihood output threshold (default 50)" << endl;
  cerr << "\t -L --length: 			length output threshold (default 100000)" << endl;
  cerr << "\t -w --wsize: 			window size (default 20)" << endl;
  cerr << "\t -x --maxerr: 			stop when encountering a window with this many ( 0/0 , 1/1 ) pairs (default 2)" << endl;
  cerr << "\t -l --lsize: 			try to join long regions closer than this (default 100000)" << endl;
  exit(1);
}

//contiguous region
class region{
public:
	int begin;
	int end;
	int matches;
	float score;	

	region() : begin(0), end(0), matches(0), score(logz){}
	region(const region& other) : begin(other.begin), end(other.end), matches(other.matches), score(other.score){}
		
	friend void swap(region& first, region& second)
    {
        swap(first.begin, second.begin);
        swap(first.end, second.end);
        swap(first.matches, second.matches);
        swap(first.score, second.score);
    }	
    region& operator=(region other)
	{
		swap(*this, other); 
		return *this;
	}
	
	region& operator+=(const region &rhs) {
		end = rhs.end;
		matches += rhs.matches;
		score +=rhs.score;
		return *this;
    }
    const region operator+(const region &other) const {
		return region(*this) += other;
    }
    
    friend bool operator<(const region& l, const region& r){ return l.score < r.score; }
    friend bool operator>(const region& l, const region& r){ return r<l; }
    friend bool operator<=(const region& l, const region& r){ return !(l > r); }
    friend bool operator>=(const region& l, const region& r){ return !(l < r); }
};

//compare regions by start position
bool chrom_compare(const region& l, const region& r){ return l.begin < r.begin; }

int ibd_main(int argc, char* argv[])
{
		
	int c;
  
	if(argc<3) usage();
	static struct option loptions[] =    {
	{"regions",1,0,'r'},	
	{"nthreads",1,0,'n'},	
	{"targets-file",1,0,'T'},
	{"thin",1,0,'h'},
	{"wsize",1,0,'w'},
	{"lsize",1,0,'l'},
	{"aftag",1,0,'a'},
	{"pairfile",1,0,'f'},
	{"error",1,0,'e'},
	{"min_freq",1,0,'m'},
	{"thresh",1,0,'M'},
	{"length",1,0,'L'},
	{"maxerr",1,0,'x'},
	{"samples",1,0,'s'},
    {"samples-file",1,0,'S'},
	{0,0,0,0}
	};

	string pfilename = "";

	int thin = 1;
	int nthreads = -1;
	int wsize = 20;
	int lsize = 100000;
	string af_tag = "";
	float thresh = 50;
	float min_len = 100000;
	string pairfile="";
	float error = 1e-3;

	float min_freq = 0;
    sample_args sargs;

    string regions = "";
    bool used_r = false;

	int maxerr = 2;
	
	while ((c = getopt_long(argc, argv, "r:h:n:w:l:a:T:L:f:e:M:m:x:s:S:",loptions,NULL)) >= 0) {  
		switch (c) {
			case 'r': regions = (optarg); used_r = true; break;
			case 'T': pfilename = (optarg); break;
			case 'h': thin = atoi(optarg); break;
			case 'n': nthreads = atoi(optarg); break;
			case 'M': thresh = atof(optarg); break;
			case 'L': min_len = atoi(optarg); break;
			case 'm': min_freq = atof(optarg); break;
			case 'e': error = atof(optarg); break;
			case 'w': wsize = atoi(optarg); break;
			case 'l': lsize = atoi(optarg); break;
			case 'a': af_tag = string(optarg) + "_"; break;
			case 'f': pairfile = (optarg); break;
			case 'x': maxerr = atoi(optarg); break;
			case 's': sargs.sample_names = (optarg); sargs.subsample = true; break;
			case 'S': sargs.sample_names = (optarg); sargs.subsample = true; sargs.sample_is_file = 1; break;
			case '?': usage();
			default: cerr << "Unknown argument:"+(string)optarg+"\n" << endl; exit(1);
		}
	}
	if(!used_r){ cerr << "-r argument is mandatory" << endl; exit(1); }
	if(optind>=argc-1) { cerr<< "No input .bcf/.vcf provided!"<<endl;exit(1);}
    if(pfilename.empty()){ cerr << "No sites file provided" << endl;exit(1); }
    
	omp_set_num_threads(nthreads); 
	#pragma omp parallel
	{
		if(omp_get_thread_num() == 0){
			if( omp_get_num_threads() != 1){
				cerr << "Using " << omp_get_num_threads() << " threads" << endl;
			}
		}
	}
	
	optind++;
	string filename = argv[optind];
					
	bcf_srs_t *sr =  bcf_sr_init() ; ///htslib synced reader.
	sr->collapse = COLLAPSE_NONE;	//matching ALTS
	sr->require_index = 1;	//indexed files

	//set regions
	if ( bcf_sr_set_regions(sr, regions.c_str(), false)<0 ){ 
		cerr << "Failed to read the regions: " <<  regions << endl; 
		return 0;
	}
	//input file
	if(!(bcf_sr_add_reader (sr, filename.c_str() ))){ 
	cerr << "Problem opening " << filename << endl; 
	bcf_sr_destroy(sr);	
	return 0;
	}
	//sites file
	if(!(bcf_sr_add_reader (sr, pfilename.c_str() ))){ 
	cerr << "Problem opening " << pfilename << endl; 
	bcf_sr_destroy(sr);	
	return 0;
	}
	//sub sample
    if(sargs.subsample){ bcf_hdr_set_samples(sr->readers[0].header, sargs.sample_names, sargs.sample_is_file); }
		
	int N = bcf_hdr_nsamples(sr->readers[0].header);	///number of samples 
	cerr << N << " samples" << endl;
	vector<string> names;
	map<string,int> name_to_id;
	set<string> unique_names;

	for(int i=0; i<N; ++i){ 
	string tmp = sr->readers[0].header->samples[i]; 
	names.push_back(tmp);
	name_to_id[tmp] = i;
	}

	//sample pairs to process
	vector< pair<string, string> > relpairs;
	if( pairfile != "" ){
		ifstream in(pairfile.c_str());
		read_pairs(in, relpairs, name_to_id);
		in.close();
	} else {
		make_pair_list(relpairs, names);
	}

	//Dont't waste time reading all samples, just required ones
	for(size_t r=0; r<relpairs.size(); ++r){ //all sample pairs
		unique_names.insert( relpairs[r].first  );
		unique_names.insert( relpairs[r].second  );
    }	  
	set<string>::iterator ufin = --unique_names.end();
	set<string>::iterator uit  = unique_names.begin();
	string sample_names="";
	for (; uit!=ufin; ++uit){
		sample_names += (*uit) + ",";
	} sample_names += (*uit);

    int sret = bcf_hdr_set_samples(sr->readers[0].header, sample_names.c_str(), 0);
    if(sret < 0){ cerr << "Failed to set samples from " << pairfile << endl; exit(1); }
    
	N = bcf_hdr_nsamples(sr->readers[0].header);	///number of samples 
	
	name_to_id.clear();
	for(int i=0; i<N; ++i){ 
		string tmp = sr->readers[0].header->samples[i]; 
		name_to_id[tmp] = i;
	}
	
	vector< vector<int> > G; 

	int count=0,sites=0,markers=0;
	bcf1_t *line, *line2;///bcf/vcf line structure.

	vector< float > af;
	vector< float > ld;
	int *gt_arr=(int *)malloc(N*2*sizeof(int)),ngt=N*2,ngt_arr=N*2;
	int *gq_arr=(int *)malloc(N*sizeof(int)),ngq=N,ngq_arr=N;
	float *af_ptr=(float *)malloc(1*sizeof(float)); int nval = 1;
	float *ld_ptr=(float *)malloc(1*sizeof(float)); 

	vector<int> position;

	vector< vector<float> > GQ;
    
	int num_sites=0,num_study=0;
	
	
	while(bcf_sr_next_line (sr)) { 	
	if( bcf_sr_has_line(sr,1) ){	//present in sites file.
			
	line2 =  bcf_sr_get_line(sr, 1);
	if(line2->n_allele == 2){	//bi-allelic
	int nmiss=0;
	int sum = 0;
	num_sites++;
	if(bcf_sr_has_line(sr,0)) { //present in the study file
	line =  bcf_sr_get_line(sr, 0);
	ngt = bcf_get_genotypes(sr->readers[0].header, line, &gt_arr, &ngt_arr);   
	if(ngt < 0){ cerr << "Bad genotypes at " << line->pos+1 << endl; exit(1); }
	 
	for(int i=0;i<2*N;i++){ 
		if(gt_arr[i]==bcf_gt_missing || (bcf_gt_allele(gt_arr[i])<0) || (bcf_gt_allele(gt_arr[i])>2)  ){
			gt_arr[i]=bcf_gt_phased(0);//inserting homref instead of missing.
		}
		sum += bcf_gt_allele(gt_arr[i]);
		++nmiss; 
	}
	++num_study;

	//only take one of every 'thin' of these sites
	if( (count++)%thin==0 ){ 

		float p;
		int ret = bcf_get_info_float(sr->readers[1].header, line2, (af_tag + "AF").c_str(), &af_ptr, &nval);
		if( ret < 0 || nval != 1 ){ cerr << (af_tag + "AF") << " read error at " << line2->rid << ":" << line->pos+1 << endl; exit(1); }
		p = af_ptr[0];
		
		if( (p < 0.5) ? ( p > min_freq ) : (1-p > min_freq) ){ 	//freq cutoff
						
			vector<int> tmp(N); 
			for(int i=0; i<N; ++i){ tmp[i] = bcf_gt_allele(gt_arr[2*i]) + bcf_gt_allele(gt_arr[2*i+1]); } 
			G.push_back(tmp);							
			af.push_back( p );
			position.push_back( line->pos+1 );
			
			//read GQ for errors otherwise use constant
			ngq = bcf_get_format_float(sr->readers[0].header, line, "GQ", &gq_arr, &ngq_arr);
			vector<float> tmpq(N);
			if( ngq < 0 ){
				for(int i=0; i<N; ++i){ tmpq[i] = error; } 
			} else {
				for(int i=0; i<N; ++i){ tmpq[i] = exp(-0.1*gq_arr[i]*logten); } 
			}
			GQ.push_back( tmpq );
			
			//read LD scores otherwise assume equilibrium
			float tl;
			int ret = bcf_get_info_float(sr->readers[1].header, line2, (af_tag + "LD").c_str(), &ld_ptr, &nval);
			if( ret<=-1 ){ 
				tl = 1; 
			} else {
				tl = ld_ptr[0];
			}
			ld.push_back(tl);
			
			++markers;
		}

	}//thin
	
	}//in both	
	}//biallelic
	}//in intersection
		if( bcf_sr_has_line(sr,1) ){ ++sites; }
	}//reader


	free(gt_arr);
	free(gq_arr);
	free(af_ptr);

	cerr << "Kept " << markers << " markers out of " << sites << " in panel." << endl;
	cerr << num_study << "/"<<num_sites<<" of study markers were in the sites file"<<endl;

	bcf_sr_destroy(sr);	
  
  vector< list< region > > all_extended_match_region(relpairs.size());
  cerr << "computing IBD sharing for " << relpairs.size() << " sample pairs." << endl;
  
  #pragma omp parallel for	
  for(size_t r=0; r<relpairs.size(); ++r){ //all sample pairs
  
	  int j1 = name_to_id[ relpairs[r].first  ]; //sample ids
	  int j2 = name_to_id[ relpairs[r].second ];	  
	 
	  list<region> matches;
	  ///count number of IBS>0 matches between j1 and j2 in each window
	  for(int i=0; i<(int)G.size(); i+=wsize){	
		  region rg;
		  rg.begin = i;
		  rg.end = i+wsize;
		  for(int ip=i; ip<std::min(i+wsize,(int)G.size()); ++ip){ 
			  rg.matches += (int)( abs(G[ip][j2] - G[ip][j1]) < 2 ); // if 0/0 && 1/1 == 2 else < 2
		  } 
		  if(i+wsize < (int)G.size() 
		  && position[i+wsize]-position[i] < lsize){ matches.push_back(rg); } //chop off last window and long windows
	  }
	  	 
	  ///contiguous regions with IBS>0 
	  for (list<region>::iterator it=matches.begin(); it!=matches.end(); ++it){

			if( it->matches == wsize ){	//found a perfect window
				region rg( *it );
				
				//window score
				float score = 0;
				for(int wp=0; wp<wsize; ++wp){ score += calc_snp_logodds(G, af, ld, (it->begin) + wp, j1, j2, GQ); }
				rg.score = score;

				//done this window, drop it from list
				//123(it = 4)567...
				//123(it = 5)67...
				it = matches.erase( it );

				while( it !=matches.end() && it->matches == wsize ){ //while we can add perfect windows
					
					//window score
					score = 0; 
					for(int wp=0; wp<wsize; ++wp){ score += calc_snp_logodds(G, af, ld, (it->begin) + wp, j1, j2, GQ); }
					it->score = score;
					
					rg += (*it);	//accumulate
					it = matches.erase( it ); //done this window, drop it from list
					
				}
				//save perfect region
				//123(it = 4)56
				//123(it = 4)(rg)56
				matches.insert(it, rg);
				--it; //undoes increment (necessary??)

			} //end perfect window
	  }
  
	  float initm = 0;
	  list<region> best_regions;
	  
	  //utility iterator, can't use prev and next
	  list<region>::iterator tmpit;	  	  
	  do{ 
		  //no matches!
		  if( matches.empty() ) break;

		  //most likely IBD region
		  list<region>::iterator mit = max_element(matches.begin(), matches.end());
		  initm = mit->score;
		  

		  region best_region = (*mit);
		  region init_region = best_region;
		  
		  //to the right of best region
		  list<region>::iterator rit = matches.erase(mit);
		  list<region>::iterator init_rit = rit;
		
		  //to the left of best region
		  tmpit = rit; --tmpit; 
		  list<region>::iterator lit = (rit != matches.begin()) ? tmpit : rit;
		  list<region>::iterator init_lit = lit;


		  region cur_region = best_region;
		  
		  //extend left from max	  
		  for(; lit != matches.begin(); --lit){ 
			  if( lit->matches <= wsize - maxerr ){ break; } //too many mismatches
			  else {
				  if(lit->score == logz){ //calculate window score if necessary
					  float score = 0; 
					  for(int wp=0; wp<wsize; ++wp){ score += calc_snp_logodds(G, af, ld, (lit->begin) + wp, j1, j2, GQ); }
					  lit->score = score;
				  }
				  //extend current region left
				  region tmp_region = (*lit);
				  tmp_region += cur_region;
				  cur_region = tmp_region;

				  if( cur_region.score > best_region.score){ best_region = cur_region; } //save region with highest score encountered
			  }
		  } 
		  
		  //extend right from max
		  cur_region = best_region;	  
		  for(; rit != matches.end(); ++rit){ 
			  if( rit->matches <= wsize - maxerr ){ break; } //too many mismatches
			  else {
				  if(rit->score == logz){ //calculate window score if necessary
					  float score = 0; 
					  for(int wp=0; wp<wsize; ++wp){ score += calc_snp_logodds(G, af, ld, (rit->begin) + wp, j1, j2, GQ); }
					  rit->score = score;
				  }
				  //extend current region right
				  cur_region += (*rit);
				  if( cur_region.score > best_region.score){best_region = cur_region; } //save region with highest score encountered
			  }
		  
		  } 

		 if( best_region.end != init_region.end ){ //addeed to RHS
			list<region>::iterator it = init_rit; 
			//delete used windows
			//123(4)567
			//123(5)67...
			while(it != matches.end() && it->begin < best_region.end){it = matches.erase(it);} 
		 }
		 
		 if( best_region.begin != init_region.begin ){ //addeed to LHS
			list<region>::iterator it = init_lit; 
			//delete used windows
			//12(3)4567
			//12(4)567 
			//1(2)4567...
			while(it != matches.end() && it->end > best_region.begin){ 
				if(it == matches.begin() ){ break; }
				else{ it = matches.erase(it); --it; }
			}
		 }
		 //save best region
		 best_regions.push_back( best_region );

	} while (initm > 0); //starting region had positive likelihood

	//attempt to join neighbouring regions 
	if( !best_regions.empty() ){
		
	//sort regions by position
	best_regions.sort(chrom_compare);
	
	bool merged;
	do{
		merged = false;

		tmpit = best_regions.end(); --tmpit;
		for( list<region>::iterator mit = best_regions.begin(); mit != tmpit; ++mit){
		
			list<region>::iterator rit = mit; ++rit; //region to the right
			
			if( position[ rit->begin ] - position[ mit->end ] < lsize){ //if regions are close
				
				//calculate score from gap
				float gap_score = 0;
				for(int wp=mit->end; wp<rit->begin; ++wp){ gap_score += calc_snp_logodds(G, af, ld, wp, j1, j2, GQ); }
				if(rit->score + gap_score > 0 ){ //closing gap increases score of combined segment
					*mit += *rit; //add regions
					mit->score += gap_score; //add gap
					rit = best_regions.erase(rit); //remove merged region
					tmpit = best_regions.end(); --tmpit;
					merged = true;
				}
				
			}
		}
		
	} while(merged); //stuff is still getting merged
	
	all_extended_match_region[r] = best_regions; //save best regions
	
	} //match list empty
	
  } //end pair list
  
  for(size_t r=0; r<all_extended_match_region.size(); ++r){
		for(list<region>::iterator it=all_extended_match_region[r].begin(); 
		it != all_extended_match_region[r].end(); ++it){
			if(it->score >= thresh && position[ it->end ] - position[ it->begin ] >= min_len){
				
			cout << relpairs[r].first << "\t" << relpairs[r].second << "\t";
			cout << position[ it->begin ] << "\t" << position[ it->end ]
			<< "\t" << it->score << "\t" << it->matches << "\t" << position[ it->end ] - position[ it->begin ] << endl;
			
			}
		} 
  }
  
  
  return 0;
}
